File: python/revokable_forwarder_example.py
#!/usr/bin/python
# -*- coding: UTF-8

# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/. */

# Authors:
# Michael Berg-Mohnicke <michael.berg@zalf.de>
#
# Maintainers:
# Currently maintained by the authors.
#
# This file is part of the util library used by models created at the Institute of
# Landscape Systems Analysis at the ZALF.
# Copyright (C: Leibniz Centre for Agricultural Landscape Research (ZALF)

import asyncio
import capnp
import os
from pathlib import Path
import socket
import sys
import time
from threading import Thread
import subprocess as sp

PATH_TO_SCRIPT_DIR = Path(os.path.realpath(__file__)).parent
PATH_TO_REPO = PATH_TO_SCRIPT_DIR.parent
if str(PATH_TO_REPO) not in sys.path:
    sys.path.insert(1, str(PATH_TO_REPO))
schema = capnp.load(str(PATH_TO_REPO / "capnp" / "revokable_forwarder.capnp"))

import capnp_async_helpers as async_helpers

# for the multithreading example we need support to for eventloops in every single thread
# this removes the old single threaded one and creates a new one for the main thread with multithreading support
capnp.reset_event_loop(ignore_errors=False, threaded=True)

class Alice(schema.Alice.Server):

    def __init__(self):
        self.bob = None
        self.carol = None

    #def do(self, msg, **kwargs): # do @0 (msg :Text);
    def do_context(self, context): # do @0 (msg :Text);
        print("@Alice::do | msg:", context.params.msg)
        print("@Alice::do | sending foo(carol) to Bob")
        if self.bob and self.carol:
            return self.bob.foo(self.carol).then(lambda _: ())

    def set_context(self, context): # set @0 (bob :Bob, carol :Carol);
        print("@Alice::set")
        self.bob = context.params.bob
        self.carol = context.params.carol


class Bob(schema.Bob.Server):

    def __init__(self):
        self.carol = None

    def do_context(self, context): # do @0 (msg :Text);
        print("@Bob::do | msg:", context.params.msg)
        if self.carol:
            print("@Bob::do | sending do(msg) to Carol")
            return self.carol.do("<Bobs DO message to Carol>").then(lambda _: ())

    def foo(self, carol, **kwargs): # foo @1 (carol :Carol);
        print("@Bob::foo")
        self.carol = carol


class Carol(schema.Carol.Server):

    def __init__(self):
        pass

    def do(self, msg, **kwargs): # do @0 (msg :Text);
        print("@Carol::do | msg:", msg)


class Forwarder(schema.Forwarder.Server):

    def __init__(self):
        pass

    def do(self, **kwargs): # do @0 ();
        pass


class Revoker(schema.Revoker.Server):

    def __init__(self):
        pass

    def do(self, **kwargs): # do @0 ();
        pass


def run_alice(s, aio=False):
    print("@run_alice")
    if aio:
        asyncio.run(async_helpers.serve_forever(None, s[2:], Alice()))
    else: 
        server = capnp.TwoPartyServer(s, bootstrap=Alice())
        capnp.wait_forever()


def run_bob(s, aio=False):
    print("@run_bob")
    if aio:
        asyncio.run(async_helpers.serve_forever(None, s[2:], Bob()))
    else:
        server = capnp.TwoPartyServer(s, bootstrap=Bob())
        capnp.wait_forever()


def run_carol(s, aio=False):
    print("@run_carol")
    if aio:
        asyncio.run(async_helpers.serve_forever(None, s[2:], Carol()))
    else:
        server = capnp.TwoPartyServer(s, bootstrap=Carol())
        capnp.wait_forever()


if __name__ == '__main__':

    config = {
        "actor": "aio",
        "use_threads": False,
        "use_asyncio": True,
    }
    if len(sys.argv) > 1:
        for arg in sys.argv[1:]:
            k, v = arg.split("=")
            config[k] = v.lower() == "true" if v.lower() in ["true", "false"] else v 
    #print(config)

    actor = config["actor"]
    threaded = actor == "aio" and config["use_threads"]
    is_asyncio = config["use_asyncio"] and not config["use_threads"]
    as1, as2 = socket.socketpair() if threaded else ("localhost:9991", "*:9991")
    bs1, bs2 = socket.socketpair() if threaded else ("localhost:9992", "*:9992")
    cs1, cs2 = socket.socketpair() if threaded else ("localhost:9993", "*:9993")

    if actor == "alice":
        run_alice(as2, is_asyncio)
    elif actor == "bob":
        run_bob(bs2, is_asyncio)
    elif actor == "carol":
        run_carol(cs2, is_asyncio)
    elif actor == "main":
        alice = capnp.TwoPartyClient(as1).bootstrap().cast_as(schema.Alice)
        bob = capnp.TwoPartyClient(bs1).bootstrap().cast_as(schema.Bob)
        carol = capnp.TwoPartyClient(cs1).bootstrap().cast_as(schema.Carol)

        print("@main | sleep for 1s")
        time.sleep(1)

        print("@main | sending do(msg) to Alice")
        alice.set(bob, carol)
        alice.do("<mains DO message to Alice>").wait()

        print("@main | sending do(msg) to Bob")
        bob.do("<mains DO message to Bob>").wait()

        time.sleep(2)
        print("@main | finished")

    elif actor == "aio":
        # start Alice
        if threaded:
            alice_thread = Thread(name="Alice", target=run_alice, args=[as2], daemon=True)
            alice_thread.start()
        else:
            alice_process = sp.Popen(["python", "python/revokable_forwarder_example.py", "actor=alice", "use_asyncio="+str(is_asyncio)])

        # start Bob
        if threaded:
            bob_thread = Thread(name="Bob", target=run_bob, args=[bs2], daemon=True)
            bob_thread.start()
        else:
            bob_process = sp.Popen(["python", "python/revokable_forwarder_example.py", "actor=bob", "use_asyncio="+str(is_asyncio)])

        # start Carol
        if threaded:
            carol_thread = Thread(name="Carol", target=run_carol, args=[cs2], daemon=True)
            carol_thread.start()
        else:
            carol_process = sp.Popen(["python", "python/revokable_forwarder_example.py", "actor=carol", "use_asyncio="+str(is_asyncio)])
        

        # wait for threads/processes to have started
        time.sleep(1)
        
        # get references to Actors
        alice = capnp.TwoPartyClient(as1).bootstrap().cast_as(schema.Alice)
        bob = capnp.TwoPartyClient(bs1).bootstrap().cast_as(schema.Bob)
        carol = capnp.TwoPartyClient(cs1).bootstrap().cast_as(schema.Carol)
        
        print("@main | sending set(bob,carol) to Alice")
        alice.set(bob, carol).wait()
        print("@main | sending do(msg) to Alice")
        alice.do("<mains DO message to Alice>").wait()

        print("@main | sending do(msg) to Bob")
        bob.do("<mains DO message to Bob>").wait()

        time.sleep(2)
        if not threaded:
            alice_process.terminate()
            bob_process.terminate()
            carol_process.terminate()
        print("@main | finished")

File: csharp/revokable_forwarder_example.cs
using System;
using Capnp.Rpc;
using Capnp.Rpc.Interception;
using System.Net;
using System.Text.Json;
using System.Threading.Tasks;
using System.Threading;
using System.Diagnostics;

namespace Ems_ocaps_paper
{
    class Alice : Schema.IAlice {
        readonly object _lock = new object();
        private Schema.IBob _bob;
        private Schema.ICarol _carol;
        private Schema.IForwarder _forwarder = new Forwarder();
        private Schema.IRevoker _revoker = new Revoker();

        // do @0 (msg :Text);
        public async Task Act(string msg, CancellationToken cancellationToken_ = default) {
            Console.WriteLine("@Alice::act | msg: " + msg);
            Console.WriteLine("@Alice::act | sending foo(carol) to Bob");
            if(_bob != null && _carol != null) {
                await _forwarder.SetActor(Proxy.Share(_revoker), cancellationToken_);
                await _revoker.SetActor(Proxy.Share(_carol), cancellationToken_);
                //await _bob.Foo(Proxy.Share(_carol), cancellationToken_);
                await _bob.Foo(Proxy.Share(_forwarder), cancellationToken_);
            }
        }

        // set @0 (bob :Bob, carol :Carol);
        public Task SetBobAndCarol(Schema.IBob bob, Schema.ICarol carol, CancellationToken cancellationToken_ = default) {
            Console.WriteLine("@Alice::setBobAndCarol");
            lock(_lock){
                if (bob != _bob) {
                    _bob?.Dispose();
                    _bob = bob;
                }
                if (carol != _carol) {
                    _carol?.Dispose();
                    _carol = carol;
                }
            }
            return Task.CompletedTask;
        }

        // revokeCarol     @1 ();
        public Task RevokeCarol(CancellationToken cancellationToken_ = default) {
            Console.WriteLine("@Alice::revokeCarol");
            return _revoker.Revoke();
        }

        public void Dispose() {
            _bob?.Dispose();
            _carol?.Dispose();
        }

        static private TcpRpcServer _server;
        static public void run(int tcpPort) {
            _server = new TcpRpcServer();
            _server.Main = new Alice();
            _server.StartAccepting(IPAddress.Any, tcpPort);
        }
    }

    class Bob : Schema.IBob {
        private readonly object _lock = new object();
        private Schema.ICarol _carol;
        private int _count = 1;

        // act @0 (msg :Text);
        public async Task Act(string msg, CancellationToken cancellationToken_ = default) {
            Console.WriteLine("@Bob::act | msg: " + msg);
            if(_carol != null) {
                Console.WriteLine("@Bob::act | sending act(msg) to Carol");
                await _carol.Act("<Bobs " + _count + ". ACT message to Carol>", cancellationToken_);
                _count++;
            }   
        }

        // foo @0 (carol :Carol);
        public Task Foo(Schema.ICarol carol, CancellationToken cancellationToken_ = default) {
            Console.WriteLine("@Bob::foo");
            lock (_lock) {
                if (carol != _carol) {
                    _carol?.Dispose();
                    _carol = carol;
                }
            }
            return Task.CompletedTask;
        }

        public void Dispose() {}

        static private TcpRpcServer _server;
        static public void run(int tcpPort) {
            _server = new TcpRpcServer();
            _server.Main = new Bob();
            _server.StartAccepting(IPAddress.Any, tcpPort);
        }
    }

    class Carol : Schema.ICarol {
        // act @0 (msg :Text);
        public Task Act(string msg, CancellationToken cancellationToken_ = default) {
            Console.WriteLine("@Carol::act | msg: " + msg);
            return Task.CompletedTask;
        }

        public void Dispose() {}

        static private TcpRpcServer _server;
        static public void run(int tcpPort) {
            _server = new TcpRpcServer();
            _server.Main = new Carol();
            _server.StartAccepting(IPAddress.Any, tcpPort);
        }
    }

    class Forwarder : Schema.IForwarder {
        private readonly object _lock = new object();
        private Schema.IActor _actor;

        // act @0 (msg :Text);
        public async Task Act(string msg, CancellationToken cancellationToken_ = default) {
            Console.WriteLine("@Forwarder::act | msg: " + msg);
            await _actor?.Act(msg, cancellationToken_);
        }

        // setActor @0 (a :Actor);
        public Task SetActor(Schema.IActor actor, CancellationToken cancellationToken_ = default) {
            Console.WriteLine("@Forwarder::setActor");
            lock (_lock) {
                if (actor != _actor) {
                    _actor?.Dispose();
                    _actor = actor;
                }
            }
            return Task.CompletedTask;
        }

        public void Dispose() {}

        static private TcpRpcServer _server;
        static public void run(int tcpPort) {
            _server = new TcpRpcServer();
            _server.Main = new Forwarder();
            _server.StartAccepting(IPAddress.Any, tcpPort);
        }
    }

    class Revoker : Schema.IRevoker {
        private readonly object _lock = new object();
        private Schema.IActor _actor;

        // act @0 (msg :Text);
        public async Task Act(string msg, CancellationToken cancellationToken_ = default) {
            Console.WriteLine("@Revoker::act | msg: " + msg);
            if (_actor != null) await _actor.Act(msg, cancellationToken_);
            else Console.WriteLine("Access to Actor has been revoked.");
        }

        // setActor @0 (a :Actor);
        public Task SetActor(Schema.IActor actor, CancellationToken cancellationToken_ = default) {
            Console.WriteLine("@Revoker::setActor");
            lock (_lock) {
                if (actor != _actor) {
                    _actor?.Dispose();
                    _actor = actor;
                }
            }
            return Task.CompletedTask;
        }

        // revoke @0 ();
        public Task Revoke(CancellationToken cancellationToken_ = default) {
            Console.WriteLine("@Revoker::revoke");
            lock (_lock) {
                _actor?.Dispose();
                _actor = null;
            }
            return Task.CompletedTask;
        }

        public void Dispose() {}

        static private TcpRpcServer _server;
        static public void run(int tcpPort) {
            _server = new TcpRpcServer();
            _server.Main = new Revoker();
            _server.StartAccepting(IPAddress.Any, tcpPort);
        }
    }

    class Program
    {
        static async Task Main(string[] args)
        {
            string actor = "aio";
            int alicePort = 9991;
            int bobPort = 9992;
            int carolPort = 9993;

            for (var i = 0; i < args.Length; i++)
            {
                try
                {
                    if (args[i].StartsWith("actor")) actor = args[i].Split('=')[1];
                }
                catch (System.Exception) { }
            }

            if(actor == "alice") Alice.run(alicePort);
            else if(actor == "bob") Bob.run(bobPort);
            else if(actor == "carol") Carol.run(carolPort);
            else if(actor == "main")
            {
                using var aliceCon = new TcpRpcClient("127.0.0.1", alicePort);
                var alice = aliceCon.GetMain<Schema.IAlice>();
                using var bobCon = new TcpRpcClient("127.0.0.1", bobPort);
                var bob = bobCon.GetMain<Schema.IBob>();
                using var carolCon = new TcpRpcClient("127.0.0.1", carolPort);
                var carol = carolCon.GetMain<Schema.ICarol>();

                Console.WriteLine("@main | sleep for 1s");
                System.Threading.Thread.Sleep(1000);

                Console.WriteLine("@main | sending setBobAndCarol(bob,carol) to Alice");
                await alice.SetBobAndCarol(Proxy.Share(bob), carol);
                Console.WriteLine("@main | sending act(msg) to Alice");
                await alice.Act("<mains ACT message to Alice>");

                Console.WriteLine("@main | sending act(msg) to Bob");
                await bob.Act("<mains 1st ACT message to Bob>");
                await bob.Act("<mains 2nd ACT message to Bob>");

                Console.WriteLine("@main | sending revokeCarol to Alice");
                await alice.RevokeCarol();

                Console.WriteLine("@main | sending act(msg) to Bob");
                await bob.Act("<mains 3rd ACT message to Bob>");

                //System.Threading.Thread.Sleep(4000);
                Console.WriteLine("@main | finished");

                //while (true) System.Threading.Thread.Sleep(1000); 
            }
        }
    }
}
